src/test/fuzz/tx_pool.cpp

```diff
@@ -239,33 +253,52 @@ FUZZ_TARGET_INIT(tx_pool_standard, initialize_tx_pool)
 
-        // Make sure ProcessNewPackage on one transaction works.
-        // The result is not guaranteed to be the same as what is returned by ATMP.
-        const auto result_package = WITH_LOCK(::cs_main,
-                                    return ProcessNewPackage(chainstate, tx_pool, {tx}, true));
-        // If something went wrong due to a package-specific policy, it might not return a
-        // validation result for the transaction.
-        if (result_package.m_state.GetResult() != PackageValidationResult::PCKG_POLICY) {
-            auto it = result_package.m_tx_results.find(tx->GetWitnessHash());
-            Assert(it != result_package.m_tx_results.end());
-            Assert(it->second.m_result_type == MempoolAcceptResult::ResultType::VALID ||
-                   it->second.m_result_type == MempoolAcceptResult::ResultType::INVALID);
+
+        bool accepted{true};
+        for (const auto& tx : package) {
+            const bool submit_standalone{
+                // A package must have at least 2 txs, so size-1 packages can only be submitted standalone
+                package.size() == 1 ||
+                fuzzed_data_provider.ConsumeBool()};
+            if (submit_standalone) {
+                const auto result_single{WITH_LOCK(::cs_main, return AcceptToMemoryPool(chainstate, tx, GetTime(), bypass_limits, /*test_accept=*/false))};
+                const bool accepted_single{result_single.m_result_type == MempoolAcceptResult::ResultType::VALID};
+                accepted &= accepted_single;
+            }
         }
 
-        const auto res = WITH_LOCK(::cs_main, return AcceptToMemoryPool(chainstate, tx, GetTime(), bypass_limits, /*test_accept=*/false));
-        const bool accepted = res.m_result_type == MempoolAcceptResult::ResultType::VALID;
+        if (package.size() >= 2) {
+            // Submit the remainder of the package.
+            // For now, every package has only one tx , so this submits the whole
+            // package again.
+            const auto result_package{WITH_LOCK(::cs_main, return ProcessNewPackage(chainstate, tx_pool, package, /*test_accept=*/false))};
+            accepted = result_package.m_state.IsValid();
+            Assert(accepted != result_package.m_state.IsInvalid());
+            // If something went wrong due to a package-specific policy, it might not return a
+            // validation result for the transaction.
+            if (result_package.m_state.GetResult() != PackageValidationResult::PCKG_POLICY) {
+                for (const auto& tx_res : result_package.m_tx_results) {
+                    Assert(tx_res.second.m_result_type == MempoolAcceptResult::ResultType::VALID ||
+                           tx_res.second.m_result_type == MempoolAcceptResult::ResultType::INVALID);
+                }
+            }
+            if (accepted) {
+                Assert(result_package.m_tx_results.size() == package.size());
```

Comment:
The assertion `result_package.m_tx_results.size() == package.size()` failed. I debug the testcase and it seems results.size() < package.size().

```
test/fuzz/tx_pool.cpp:286 tx_pool_standard_fuzz_target: Assertion `result_package.m_tx_results.size() == package.size()' failed.
==3398073== ERROR: libFuzzer: deadly signal
    #0 0x55e1fcaec680 in __sanitizer_print_stack_trace (/home/ubuntu/bitcoin/src/test/fuzz/fuzz+0xfd8680)
    #1 0x55e1fca98988 in fuzzer::PrintStackTrace() (/home/ubuntu/bitcoin/src/test/fuzz/fuzz+0xf84988)
    #2 0x55e1fca7dad3 in fuzzer::Fuzzer::CrashCallback() (/home/ubuntu/bitcoin/src/test/fuzz/fuzz+0xf69ad3)
    #3 0x7ff789c3741f  (/lib/x86_64-linux-gnu/libpthread.so.0+0x1441f)
    #4 0x7ff78989d00a in __libc_signal_restore_set /build/glibc-SzIz7B/glibc-2.31/signal/../sysdeps/unix/sysv/linux/internal-signals.h:86:3
    #5 0x7ff78989d00a in raise /build/glibc-SzIz7B/glibc-2.31/signal/../sysdeps/unix/sysv/linux/raise.c:48:3
    #6 0x7ff78987c858 in abort /build/glibc-SzIz7B/glibc-2.31/stdlib/abort.c:79:7
    #7 0x55e1fee2dc3d in assertion_fail(char const*, int, char const*, char const*) src/util/check.cpp:13:5
    #8 0x55e1fd04bc02 in bool&& inline_assertion_check<true, bool>(bool&&, char const*, int, char const*, char const*) src/./util/check.h:67:13
    #9 0x55e1fd2f9f0b in (anonymous namespace)::tx_pool_standard_fuzz_target(Span<unsigned char const>) src/test/fuzz/tx_pool.cpp:286:17
```

Reply:
Ok, that is probably my fault. I wrote the fuzz code without reading the validation code. I should read more about the interface and the meaning of valid/invalid in the context of packages. Apparently we can't deduce any meaningful information useful for fuzzing from the result. So for now I'd suggest to remove the failing Asserts.

Moreover, the "tracking" code seems not prepared for packages. I think it can be fixed by tracking the history of added/removed transactions chronologically. (That is, switching the two std::set into one std::vector)

See this diff:

```diff
diff --git a/src/test/fuzz/tx_pool.cpp b/src/test/fuzz/tx_pool.cpp
index c38e3420f6..b01b4f978a 100644
--- a/src/test/fuzz/tx_pool.cpp
+++ b/src/test/fuzz/tx_pool.cpp
@@ -70,6 +70,24 @@ struct TransactionsDelta final : public CValidationInterface {
     }
 };
 
+struct TransactionsHistory final : public CValidationInterface {
+    std::vector<std::tuple<bool, CTransactionRef>>& m_history;
+
+    explicit TransactionsHistory(std::vector<std::tuple<bool, CTransactionRef>>& h)
+        : m_history{h} {}
+
+    void TransactionAddedToMempool(const CTransactionRef& tx, uint64_t /* mempool_sequence */) override
+    {
+        m_history.emplace_back(true, tx);
+    }
+
+    void TransactionRemovedFromMempool(const CTransactionRef& tx, MemPoolRemovalReason reason, uint64_t /* mempool_sequence */) override
+    {
+        if (reason != MemPoolRemovalReason::SIZELIMIT)
+            m_history.emplace_back(false, tx);
+    }
+};
+
 void SetMempoolConstraints(ArgsManager& args, FuzzedDataProvider& fuzzed_data_provider)
 {
     args.ForceSetArg("-limitancestorcount",
@@ -247,9 +265,8 @@ FUZZ_TARGET_INIT(tx_pool_standard, initialize_tx_pool)
         }
 
         // Remember all removed and added transactions
-        std::set<CTransactionRef> removed;
-        std::set<CTransactionRef> added;
-        auto txr = std::make_shared<TransactionsDelta>(removed, added);
+        std::vector<std::tuple<bool, CTransactionRef>> history;
+        auto txr = std::make_shared<TransactionsHistory>(history);
         RegisterSharedValidationInterface(txr);
         const bool bypass_limits = fuzzed_data_provider.ConsumeBool();
 
@@ -283,7 +300,6 @@ FUZZ_TARGET_INIT(tx_pool_standard, initialize_tx_pool)
                 }
             }
             if (accepted) {
-                Assert(result_package.m_tx_results.size() == package.size());
                 for (const auto& tx_res : result_package.m_tx_results) {
                     Assert(tx_res.second.m_result_type == MempoolAcceptResult::ResultType::VALID);
                 }
@@ -291,15 +307,6 @@ FUZZ_TARGET_INIT(tx_pool_standard, initialize_tx_pool)
         }
         SyncWithValidationInterfaceQueue();
         UnregisterSharedValidationInterface(txr);
-        Assert(accepted != added.empty());
-        if (accepted) {
-            Assert(Package(added.begin(), added.end()) == package);
-        } else {
-            // Do not consider rejected transactions removed
-            for (const auto& tx : package) {
-                removed.erase(tx);
-            }
-        }
 
         // Helper to insert spent and created outpoints of a tx into collections
         using Sets = std::vector<std::reference_wrapper<std::set<COutPoint>>>;
@@ -317,16 +324,17 @@ FUZZ_TARGET_INIT(tx_pool_standard, initialize_tx_pool)
         };
         // Add created outpoints, remove spent outpoints
         {
+            for (const auto& [added,tx]:history){
             // Outpoints that no longer exist at all
             std::set<COutPoint> consumed_erased;
             // Outpoints that no longer count toward the total supply
             std::set<COutPoint> consumed_supply;
-            for (const auto& removed_tx : removed) {
-                insert_tx(/*created_by_tx=*/{consumed_erased}, /*consumed_by_tx=*/{outpoints_supply}, /*tx=*/*removed_tx);
-            }
-            for (const auto& added_tx : added) {
-                insert_tx(/*created_by_tx=*/{outpoints_supply, outpoints_rbf}, /*consumed_by_tx=*/{consumed_supply}, /*tx=*/*added_tx);
+            if (added) {
+                insert_tx(/*created_by_tx=*/{outpoints_supply, outpoints_rbf}, /*consumed_by_tx=*/{consumed_supply}, /*tx=*/*tx);
+            } else {
+                insert_tx(/*created_by_tx=*/{consumed_erased}, /*consumed_by_tx=*/{outpoints_supply}, /*tx=*/*tx);
             }
+
             for (const auto& p : consumed_erased) {
                 Assert(outpoints_supply.erase(p) == 1);
                 Assert(outpoints_rbf.erase(p) == 1);
@@ -334,6 +342,7 @@ FUZZ_TARGET_INIT(tx_pool_standard, initialize_tx_pool)
             for (const auto& p : consumed_supply) {
                 Assert(outpoints_supply.erase(p) == 1);
             }
+            }
         }
     }
     Finish(fuzzed_data_provider, tx_pool, chainstate);
```

However, now I am running into an assert in validation, which might be an actual bug :smiling_face_with_tear:

Reply:
Sorry for the late response, your diff is applied in https://github.com/bitcoin/bitcoin/commit/08e5172fbc5c4f5032c10dde95d463e33b644e2a, with some formatting changes by clangd.

Reply:
The crash happens when one of the tx finally get passed into `CheckInputsFromMempoolAndCache`, it assume that the tx is either a UTXO from mempool or UTXO set. The assert fails because `coinFromUTXOSet` is actually `coinEmpty`.